kliff/trainer/torch_trainer_utils/lightning_checkpoints.py

```diff
@@ -0,0 +1,92 @@
+import os
+
+import dill
+import pytorch_lightning as pl
+import torch
+
+from kliff.dataset import Dataset
+
+
+class SaveModelCallback(pl.Callback):
+    """
+    Callback to save the model at the end of each epoch. The model is saved in the ckpt_dir with the name
+    "last_model.pth". The best model is saved with the name "best_model.pth". The model is saved every
+    ckpt_interval epochs with the name "epoch_{epoch}.pth".
+    """
+
+    def __init__(self, ckpt_dir, ckpt_interval=100):
+        super().__init__()
+        self.ckpt_dir = ckpt_dir
+        self.best_val_loss = float("inf")
+        self.ckpt_interval = ckpt_interval
+        os.makedirs(self.ckpt_dir, exist_ok=True)
+
+    def on_validation_epoch_end(
+        self, trainer: pl.Trainer, pl_module: pl.LightningModule
+    ):
+        # Save the last model
+        last_save_path = os.path.join(self.ckpt_dir, "last_model.pth")
+        torch.save(pl_module.state_dict(), last_save_path)
+
+        # Save the best model
+        if trainer.callback_metrics.get("val_loss") < self.best_val_loss:
+            self.best_val_loss = trainer.callback_metrics["val_loss"]
+            best_save_path = os.path.join(self.ckpt_dir, "best_model.pth")
+            torch.save(pl_module.state_dict(), best_save_path)
+
+        # Save the model every ckpt_interval epochs
+        if pl_module.current_epoch % self.ckpt_interval == 0:
+            epoch_save_path = os.path.join(
+                self.ckpt_dir, f"epoch_{pl_module.current_epoch}.pth"
+            )
+            torch.save(pl_module.state_dict(), epoch_save_path)
+
+        # save the trainer checkpoint as well
+        trainer.save_checkpoint(os.path.join(self.ckpt_dir, "trainer_checkpoint.ckpt"))
+
+
+class LossTrajectoryCallback(pl.Callback):
```

Comment:
None of the existing loggers quite solve this problem. But I think it is a good idea to write own logger for this using lightning logger class. Will write one and try.

Reply:
What functionality do you want to have but is not provided by the default loggers?

Reply:
This feature needs to be more fleshed out, which I would do after discussion with Josh (as he is the one who needs it). His requirement is to track atomwise losses in _both training and validation_ dataset and use it as a measure in some form of active learning where most pathological configurations are given different weightage (that is how I understood it). 
The biggest limitations I was thinking of at the time was 
i. binary format, for longer training epochs, or larger configurations, we might switch to pkl, or npy binary format later on, as otherwise the file sizes balloon up a lot.
ii. logging the atomwise losses, along with indexes. At present I keep the validation dataset unshuffled and save the losses, but ideally I would like to save it basex on `idx` from the trainer. I could not figure out a way to use existing logger like CSVLogger to log this information.

I tried writing a custom logger yesterday, but now I feel that using a CSV or custom logger might result in lot of twisted code, and callback might be cleanest solution.

I think overall problem is that it is somewhat underspecified problem that I am keeping here for making sure it exists in the API what we can improve upon as needed.